### PR TITLE
Scripts for rule inference using LLM

### PR DESCRIPTION
**Pull request by danieltrt**
This PR includes a basic python script to infer simple piranha rules from code examples.
I had to close the previous PR because it wasn't syncing with the remote branch

### CODE REVIEW

**Review by ketkarameya — COMMENTED** [review]
Left some comments to improve the readme file. Else it looks good to me.

File: experimental/rule_inference/base_prompt.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            + BasePrompt.input_template.format(**kwargs)
+        )
+        return [{"role": "user", "content": formatted}]
+
+    @staticmethod
+    def _get_examples(path_to_examples_rules):
+        task_examples = ""
+        for root, dirs, files in os.walk(path_to_examples_rules):
+            for file in files:
+                if file.endswith("rules.toml"):
```
Comment (ketkarameya) [inline experimental/rule_inference/base_prompt.py:119]: Does this imply that currently `PiranhaAgent` does not understand the concept of edges?

Comment (danieltrt) [reply experimental/rule_inference/base_prompt.py:119]: Yes

Comment (danieltrt) [reply experimental/rule_inference/base_prompt.py:119]: (for now)

File: experimental/rule_inference/README.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+2. A prompt is formulated using the task explanation, examples, and input template.
+3. This prompt is fed to the GPT-4 model, which generates a TOML file encapsulating the refactoring rules.
+4. The generated TOML file is output to the console.
+
+## Files
+
+- 'main.py': The main script to run PiranhaAgent.
+- 'base_prompt.py': Contains the prompt template, and helper functions build specific prompts.
+
+We also feed the model the cleanup rules as examples, which can be found in the [src/cleanup_rules](../../src/cleanup_rules) directory.
```
[the remark this answers is not in the document]
Comment (danieltrt) [reply experimental/rule_inference/README.md:52]: Added instructions for the demo

File: experimental/rule_inference/README.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+## Requirements
+
+The following are the prerequisites to use PiranhaAgent:
+
+- Python 3.7 or higher
+- OpenAI Python library
+- Tree-sitter Python library
+- Tree-sitter-languages Python library
+## Install
```
Comment (ketkarameya) [inline experimental/rule_inference/README.md]: Ow. I meant you have to fill in the holes and put text like that

Comment (ketkarameya) [reply experimental/rule_inference/README.md]: can we have a simple bash snippet to "install" this tooling?

**Review by ketkarameya — APPROVED** [review]
LGTM

**Review by ketkarameya — APPROVED** [review]
LGTM

File: experimental/rule_inference/README.md (hunk 3 above)
Comment (danieltrt) [reply experimental/rule_inference/README.md]: ah I see! like install.sh?

**Pull request merged**